lib/browser/api/power-monitor.ts

```diff
@@ -1,38 +1,46 @@
-'use strict';
+import { EventEmitter } from 'events';
+import { app } from 'electron';
 
-import { createLazyInstance } from '../utils';
+const { createPowerMonitor, getSystemIdleState, getSystemIdleTime } = process.electronBinding('power_monitor');
 
-const { EventEmitter } = require('events');
-const { createPowerMonitor, PowerMonitor } = process.electronBinding('power_monitor');
+class PowerMonitor extends EventEmitter {
+  constructor () {
+    super();
+    // Don't start the event source until both a) the app is ready and b)
+    // there's a listener registered for a powerMonitor event.
+    this.once('newListener', () => {
+      app.whenReady().then(() => {
+        const pm = createPowerMonitor();
+        pm.emit = this.emit.bind(this);
 
-// PowerMonitor is an EventEmitter.
-Object.setPrototypeOf(PowerMonitor.prototype, EventEmitter.prototype);
+        if (process.platform === 'linux') {
+          // On Linux, we inhibit shutdown in order to give the app a chance to
+          // decide whether or not it wants to prevent the shutdown. We don't
+          // inhibit the shutdown event unless there's a listener for it. This
+          // keeps the C++ code informed about whether there are any listeners.
+          pm.setListeningForShutdown(this.listenerCount('shutdown') > 0);
+          this.on('newListener', (event) => {
+            if (event === 'shutdown') {
+              pm.setListeningForShutdown(this.listenerCount('shutdown') + 1 > 0);
+            }
+          });
+          this.on('removeListener', (event) => {
+            if (event === 'shutdown') {
+              pm.setListeningForShutdown(this.listenerCount('shutdown') - 1 > 0);
```

Comment:
if there is some usage like

```
pm.on('shutdown', l1);
pm.on('shutdown', l2);
pm.off('shutdown', l1);  // listenerCount will be 1 but the call will 
                         // be `pm.setListeningForShutdown(false)` 
                         // and reset the shutdown handler, also unblocks
                        // shutdown while there is an active listener
```

Reply:
hm, i don't think that's true—adding a listener increases listener count by 1, removing a listener decreases it by 1. but the change in count doesn't happen until after the `removeListener` / `newListener` event is fired, so this `- 1` accounts for that.

Reply:
oops, missed this in the docs:

> The 'removeListener' event is emitted after the listener is removed.

will fix.

Reply:
https://repl.it/repls/ExcitingColorlessDataset yeah that seems to be true for newListener, but removeListener seems to do the opposite.